File: examples/comparisons/8.queens.recursive.pl
#!/usr/bin/perl

use strict;
use warnings;

sub test_position {
  my($new, $rest) = @_;

  for(my $i = 0; $i < @$rest; $i++) {
    if($new == $rest->[$i]) { return; }
    if($new == $rest->[$i]+(@$rest-$i)) { return; }
    if($new == $rest->[$i]-(@$rest-$i)) { return; }
  }
  return 1;
}

sub f_8_queens {
  f_8_queens_helper(8, []);
}

sub f_8_queens_helper {
  my($n, $sofar) = @_;

  if($n == 0) {
      for my $c (@$sofar) {
          print $c, " ";
      }
      print "\n";
      return;
  }

  for(my $i = 0; $i < 8; $i++) {
    if(test_position($i, $sofar)) {
      f_8_queens_helper($n-1, [@$sofar, $i]);
    }
  }
}

f_8_queens();

File: examples/comparisons/8.queens.recursive.py
#!/usr/bin/python

def test_position(new, rest):
  l = len(rest)
  for i in range(l):
    if new == rest[i] or new == rest[i]+(l-i) or new == rest[i]-(l-i):
      return False
  return True

def f_8_queens():
  f_8_queens_helper(8, []);

def f_8_queens_helper(n, sofar):
  if n == 0:
    print sofar
    return

  for i in range(8):
    if test_position(i, sofar):
      f_8_queens_helper(n-1, sofar + [i]);

f_8_queens()
